### PR TITLE
Fix #2075: Add documentation explaining ChatCompletion format internals for fine-tuning

### LINKED ISSUE #2075: How does openAI framework converts the ChatComplete (training data ) into model ready format (SFT)
I have a question regarding the internal processing pipeline for Supervised Fine‑Tuning (SFT) when using the ChatCompletion-style JSON format for training data.

{
  "messages": [
    {
      "role": "system",
      "content": "You are an ..."
    },
    {
      "role": "user",
      "content": "Please tell me about the paper \"K205...\......""
    },
    {
      "role": "assistant",
      "content": "The paper titled \"K205R specific nanobody...\""
    }
  ]
}

How does the framework internally transform the ChatCompletion training data format into a model ready input. And how is the loss computed

### PR DESCRIPTION
**Pull request by jalateras**
## Summary

This PR addresses issue #2075 by adding comprehensive documentation that explains how OpenAI's framework internally transforms ChatCompletion training data format into model-ready input for supervised fine-tuning (SFT).

## What Changed

Added a new section "Understanding the Internals: How ChatCompletion Format is Transformed for Fine-Tuning" to the `How_to_finetune_chat_models.ipynb` notebook that includes:

- **Token-Level Transformation**: Detailed explanation of how messages with roles are converted into token sequences with special delimiter tokens
- **Loss Computation**: Documentation of selective loss calculation on assistant tokens only, with system/user tokens masked
- **Practical Examples**: Code demonstrating token counting and distribution analysis
- **Best Practices**: Recommendations based on understanding the internal processing

## Why This Change

Users were asking about the internal mechanics of how the ChatCompletion JSON format is processed during fine-tuning. The existing documentation showed how to use the API but didn't explain what happens behind the scenes. This knowledge helps users:

- Optimize their training data for better efficiency
- Understand why loss is only computed on assistant messages
- Make informed decisions about token economics
- Design better training datasets

## Testing

- Added working code examples that demonstrate token counting
- Included tiktoken dependency for token analysis
- All code cells are functional and demonstrate the concepts

Fixes #2075

### CODE REVIEW

**Comment by jalateras** [comment]
commits are interleaved with another PR. Will close this one and open a new one

**Comment by CainKilled** [comment]
please add automations for checks by open ai pushes autoupdating all of the
open ai instantiations and forksz a feature could be something that would
send notices to forks of all changes and specify and flag pushes which
require enhanced monitoring by forked owners.  vs mandatory pushinff
features. in some caees like mine. id rather pull the changes I want and
leave the imoned I dontz monitoring
